hypothesis/h
# Vendor in `pyramid_authsanity`

This PR attempts to vendor in the `pyramid_authsanity` library and adapt it down to an identity based policy.

This involved:

 * Copy pasting chunks of it into our code and then refactoring them
 * Removing paths we don't use in that code
 * Eliminating interfaces we don't need 
 * Modifying the services we are left with to be more convenient

### What's happened in this PR

The 3rd party module `pyramid_authsanity` provides an authentication policy (Pyramid 1.x style) and these are being deprecated. Before we can update it, we therefore need to incorporate a version of it into our code base.

The approach taken here was to first copy paste parts of it in and then refactor them over and over removing the parts we don't use, and collapsing if statements when we know we only have one side of the if. In particular there are three areas around which `pyramid_authsanity` is more flexible than we need:

 * It has auth "sources" which can be swapped out via an interface (we always use cookies)
 * It has auth "services" which can be swapped out via an interface (we always use our own ticket interface)
 * Less important, it allows for the presence or absence of sessions (we are always configured with a session)
 
As we only have one "source" and one "service" these have been merged into a combined "AuthCookieService" which does both parts. `pyramid_authsanity` was mostly concerned with connecting the two things, but we don't need the distinction.

The rest of the policy is now in `h.auth.cookie_authentication_policy.CookieAuthenticationPolicy`. It's not in with the rest of the policies because I intend to split them up. The location isn't good, but it's better to have it separeate.

These two classes basically do everything we need.

### Review notes

I've left in excruciating detail the unvarnished path I took to factor in this code if you care to look. If you don't I'd suggest you view this as new code + a refactor.

* New code: Added `h.auth.cookie_authentication_policy.CookieAuthenticationPolicy` in place of `pyramid_authsanity`
* Refactor: Removed `AuthTicketService` and replaced it with `AuthCookieService` which manages cookies too

These aren't easily separable without inlining and writing tests for a huge amount more code than we've ended up with. To split this up into a PR which first introduced everything `pyramid_authsanity` did, in the way it did it, would result in a larger PR than this.

### Testing notes

The cookie policy is used when using the UI for `h`. To test this try:

 * `make services dev`
 * http://localhost:5000
 * Login as the `devdata_user`
 * Refresh the page - you should still be logged in
 * Log out - you should be logged out
 * Refresh the page - you should still be logged out
 * Login, delete the "auth" cookie, and refresh - you should be logged out
 * Login, edit the "auth" cookie value with an extra char, and refresh - you should be logged out
